src/Lean/Elab/MutualDef.lean

```diff
@@ -292,7 +292,32 @@ private def expandWhereStructInst : Macro
         `(structInstField|$id:ident := $val)
       | stx@`(letIdDecl|_ $_* $[: $_]? := $_) => Macro.throwErrorAt stx "'_' is not allowed here"
       | _ => Macro.throwUnsupported
-    let body ← `(structInst| { $structInstFields,* })
+
+    let startOfStructureTkInfo : SourceInfo :=
+      match whereTk.getPos? with
+      | some pos => .synthetic pos ⟨pos.byteIdx + 1⟩ true
+      | none => .none
+    let startOfStructureTk : Syntax := .atom startOfStructureTkInfo "{"
+
+    -- Position the closing `}` at the end of the trailing whitespace of `where $[$_:letDecl];*`.
+    -- We need an accurate range of the generated structure instance for structure field completion.
+    let structureStxTailInfo :=
+      whereStx[1].getTailInfo?
+      <|> whereStx[0].getTailInfo?
+    let endOfStructureTkInfo : SourceInfo :=
+      match structureStxTailInfo with
+      | some (SourceInfo.original _ _ trailing _) =>
+        let tokenPos := trailing.str.prev trailing.stopPos
+        let tokenEndPos := trailing.stopPos
+        .synthetic tokenPos tokenEndPos true
+      | _ => .none
+    let endOfStructureTk : Syntax := .atom endOfStructureTkInfo "}"
+
+    let body ← `(structInst| {%$startOfStructureTk $structInstFields,* }%$endOfStructureTk)
+    let body := body.raw.setInfo <|
+      match startOfStructureTkInfo.getPos?, endOfStructureTkInfo.getTailPos? with
+      | some startPos, some endPos => .synthetic startPos endPos true
+      | _, _ => .none
```

Comment:
As discussed this step should be sufficient and it may be better not to touch the infos of the atoms. Should also come with a comment why this is needed.